.github/workflows/ci.yml

```diff
@@ -351,3 +362,24 @@ jobs:
               | tee -a $GITHUB_STEP_SUMMARY >&2
             exit 1
           fi
+
+  generate_cache:
+    runs-on: ubuntu-latest
+    name: Generate cache
+    steps:
+      - uses: actions/checkout@f43a0e5ff2bd294095638e18286ca9a3d1956744 # v3.6.0
+
+      - uses: actions/cache@v3
+        with:
+          path: |
+            ~/.cargo/registry/index/
+            ~/.cargo/registry/cache/
+          key: ${{ runner.os }}-cargo-test-CI2-${{ hashFiles('**/Cargo.toml') }}
```

Comment:
I will remove the `-test-CI2-` when the PR going to merge.

Reply:
If we're just caching global Cargo files, is including the hashes of `Cargo.toml` files in the key unnecessary? Presumably the final version of this key could just be `${{ runner.os }}-cargo`?

Reply:
I believe it is necessary because we need to regenerate the cache whenever the `Cargo.toml` has been modified. Here's an example of the final version of this key: `Linux-cargo-a7354d5361cd9a3c83f18de0f397f087d1b2569d3df9d9b4f7489b77d7559185`.

Reply:
You mean because a modification to `Cargo.toml` could introduce new dependencies?

Reply:
Indeed, when we modify the `Cargo.toml` file, it typically involves incorporating new crates or adjusting the version dependencies of existing ones. In such cases, it is advisable to regenerate the cache.

Reply:
Gotcha. In that case, could you add the following? I realized that while this job probably downloads all of our dependencies in practice (`cargo manifest`'s dependencies are likely a superset of ours), that's not guaranteed to always hold.

```shell
# Ensure all crate dependencies are downloaded. There is no support
# in Cargo for downloading dependencies without also building the
# current crate [1], so we check all packages with --tests as a workaround.
# We may get errors other than network errors (ie, build errors), so we
# discard stderr. This entire job is best-effort since it's just a performance
# optimization, so discarding stderr is fine.
#
# [1] https://stackoverflow.com/a/42139535/836390
cargo check --workspace --tests >/dev/null 2>/dev/null
```

Reply:
Certainly, should it be inserted before the `cargo metadata` command within the `generate cache` section?

Reply:
Sure, and actually, while we're at it, I think that we can just adjust that comment to document everything:

```shell
# Ensure all dependencies are downloaded - both for our crates and for tools
# we use in CI. We don't care about these tools succeeding for two reasons:
# First, this entire job is best-effort since it's just a performance optimization.
# Second, there may be failures due to issues other than failing to download
# dependencies (e.g., `cargo metadata` called with a malformed `Cargo.toml`,
# build failure in our own crate or in dependencies, etc). For those reasons,
# we discard stderr and ignore status codes.
#
# For downloading our crates' dependencies in particular, note that there is
# no support for doing this directly [1], so we just check all crates using --tests.
#
# [1]  https://stackoverflow.com/a/42139535/836390
cargo check --workspace --tests >/dev/null 2>/dev/null || true
cargo manifest >/dev/null 2>/dev/null || true
```

Let's also not set `-eo pipefail`. That technically makes the first `|| true` redundant, but I'm 80% sure the second one is necessary (I assume that GitHub actions treat the status of the final command as the status of the entire job), so we may as well have both for symmetry.

Reply:
I see, I think we can use `&>` for redirection, as it will redirect all the output to `/dev/null`.

Reply:
I'm not familiar with the `cargo manifest` command and couldn't locate its manual. Could you please assist me in comprehending it?

Reply:
Sorry, typo on my part - I meant `cargo metadata`. I've updated the original comment.